### PR TITLE
Add operator GROUPING SETS

### PR DESCRIPTION
**Pull request by MaxTheHuman**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by l1t1** [comment]
nice job

**Comment by akuzm** [comment]
The tests are missing, did you forget `git add`?

File: src/Processors/QueryPlan/AggregatingStep.cpp [hunk 1]
```diff
@@ -53,108 +57,149 @@ void AggregatingStep::transformPipeline(QueryPipeline & pipeline, const BuildQue
...
+            params.keys_size = params.keys.size();
+            transform_params_ptr = std::make_shared<AggregatingTransformParams>(params, final);
+
+            aggregating_transforms.push_back(std::make_shared<AggregatingTransform>(pipeline.getHeader(), transform_params_ptr));
+
+            if (grouping_sets_number == 1)
+                break;
+
+            size_t first_column_to_add = (i == 0 ? 1 : 0);
+            auto adding_column_action = ActionsDAG::makeAddingColumnActions(pipeline.getHeader().getByPosition(params.keys_vector[first_column_to_add][0]));
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/AggregatingStep.cpp:83]: That's suspicious.
What if we have two sets with the same prefix?
For `(a, b), (a, c)` we add const column `a` for both, replacing normal column with default one.

File: src/Processors/QueryPlan/AggregatingStep.cpp [hunk 2]
```diff
@@ -53,108 +57,149 @@ void AggregatingStep::transformPipeline(QueryPipeline & pipeline, const BuildQue
...
+                if (j == i || j == first_column_to_add) continue;
+                auto helper_adding_column_action = ActionsDAG::makeAddingColumnActions(pipeline.getHeader().getByPosition(params.keys_vector[j][0]));
+                adding_column_action = ActionsDAG::merge(std::move(*helper_adding_column_action), std::move(*adding_column_action));
+            }
+            auto actions = std::make_shared<ExpressionActions>(adding_column_action, ExpressionActionsSettings{});
+
+            expression_transforms.push_back(std::make_shared<ExpressionTransform>(aggregating_transforms[i]->getOutputs().front().getHeader(), actions));
+
+        }
+        pipeline.addParallelTransforms(std::move(aggregating_transforms));
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/AggregatingStep.cpp:95]: The same could be done with `addSimpleTransform`, cause aggregating one input and one output port.
But maybe this method will be useful later, when we remove `pipeline.resize(1);`

**Review by KochetovNicolai — COMMENTED** [review]
So far so good.

Also, we need to remove all debug output

**Comment by taylor12805** [comment]
I try to reproduce the test case from website: https://oracle-base.com/articles/misc/rollup-cube-grouping-functions-and-grouping-sets

`
CREATE TABLE default.dimension_tab (
  `fact_1_id` Int32,
  `fact_2_id` Int32,
  `fact_3_id` Int32,
  `fact_4_id` Int32,
  `sales_value` Decimal(9, 2)
) ENGINE = MergeTree
ORDER BY
  tuple() SETTINGS index_granularity = 8192
`

`
INSERT INTO default.dimension_tab
SELECT rand32(number) % 2 + 1 AS fact_1_id,
       rand32(number) % 5 + 1 AS fact_2_id,
       rand32(number) % 10 + 1 AS fact_3_id,
       rand32(number) % 10 + 1 AS fact_4_id,
       rand32(number) % 100 AS sales_value
FROM  system.numbers limit 1000
`

`
SELECT fact_1_id,
       fact_2_id,
       fact_3_id,
       SUM(sales_value) AS sales_value
FROM default.dimension_tab
GROUP BY GROUPING SETS((fact_1_id, fact_2_id), (fact_1_id, fact_3_id))
ORDER BY fact_1_id, fact_2_id, fact_3_id
`

In Oracle, the result is 
`
FACT_1_ID  FACT_2_ID  FACT_3_ID SALES_VALUE GROUPING_ID
---------- ---------- ---------- ----------- -----------
         1          1                4363.55           1
         1          2                4794.76           1
         1          3                4718.25           1
         1          4                5387.45           1
         1          5                5027.34           1
         1                     1      2737.4           2
         1                     2     1854.29           2
         1                     3     2090.96           2
         1                     4     2605.17           2
         1                     5     2590.93           2
         1                     6      2506.9           2
         1                     7     1839.85           2
         1                     8     2953.04           2
         1                     9     2778.75           2
         1                    10     2334.06           2
         2          1                5652.84           1
         2          2                4583.02           1
         2          3                5555.77           1
         2          4                5936.67           1
         2          5                4508.74           1
         2                     1     3512.69           2
         2                     2     2847.94           2
         2                     3      2972.5           2
         2                     4     2534.06           2
         2                     5     3115.99           2
         2                     6     2775.85           2
         2                     7     2208.19           2
         2                     8     2358.55           2
         2                     9     1884.11           2
         2                    10     2027.16           2
`

The result in this implementation:

`
FACT_1_ID  FACT_2_ID  FACT_3_ID SALES_VALUE GROUPING_ID
---------- ---------- ---------- ----------- -----------
      0	        0      	1	4820
      0	        0      	2	5222
      0      	0      	3	5130
      0	        0	4	5426
      0      	0	5	4782
      0       	0	6	4115
      0       	0	7	5752
      0	        0	8	5385
      0      	0	9	4772
      0      	0	10	5533
      0      	1	0	8935
      0      	2	0	10974
      0      	3	0	10515
      0      	4	0	10198
      0       	5	0	10315
      1       	0	0	25256
      2      	0	0	25681
`

The result of this implementation is quite different from the result in Oracle. I suppose there is some issue with this implementation. Correct me if I am wrong

**Comment by taylor12805** [comment]
@KochetovNicolai  Hi, KochetovNicolai, I'm trying to fix grouping sets multiple elements related issue, i wonder if we should support no parenthesis 
 element inside grouping sets like: GROUPING SETS(A, (A, B), (A, C)) or should only contains  elements inside parenthesis like: GROUPING SETS((A), (A, B), (A, C))

**Comment by KochetovNicolai** [comment]
@taylor12805 Hi!
I think we need to support all the cases. Query with `group by grouping sets (x, (x), (y, z), ())` works in Postrges, so I think our syntax should also support it.

> I'm trying to fix grouping sets multiple elements related issue

Do you need some help with it? I think I may also finish this pr, but I would really appreciate any assistance from you.

**Comment by taylor12805** [comment]
> @taylor12805 Hi!
> I think we need to support all the cases. Query with `group by grouping sets (x, (x), (y, z), ())` works in Postrges, so I think our syntax should also support it.
> 
> > I'm trying to fix grouping sets multiple elements related issue
> 
> Do you need some help with it? I think I may also finish this pr, but I would really appreciate any assistance from you.

@KochetovNicolai Hi!
I am just fixed the issue of multiple grouping sets not working correctly in AggregatingStep. There are still some remaining work in TreeRewriter part. How is urs? Let me know anywhere I can give a hand

**Comment by sensatyaki** [comment]
What code changes have been done to handle the empty set with () parenthesis, at the parsing level? For example ((A),()).

**Comment by novikd** [comment]
GROUPING SETS are added in #26869

**Pull request closed without merging**